File: 2016/Day_1/day1.go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
)

type Coordinate struct {
	x, y int
}

func getDirection(facing string, turning string) string {
	var new_direction string

	compass := []byte{'W', 'N', 'E', 'S'}
	facing_index := bytes.IndexAny([]byte(compass), facing)

	if turning == "L" {
		if facing_index-1 >= 0 {
			new_direction = string(compass[facing_index-1])
		} else {
			new_direction = string(compass[len(compass)-1])
		}
	} else {
		if facing_index+1 < len(compass) {
			new_direction = string(compass[facing_index+1])
		} else {
			new_direction = string(compass[0])
		}
	}
	return new_direction
}

func blocksAway(x, y int) int {
	if x < 0 && y >= 0 {
		return -x + y
	} else if x >= 0 && y < 0 {
		return x + -y
	} else if x < 0 && y < 0 {
		return -x + -y
	} else {
		return x + y
	}
}

func part1(directions []string) int {
	hq_location := Coordinate{0, 0}
	facing := "N"
	for _, direc := range directions {
		turn := string(direc[0])
		distance, err := strconv.Atoi(string(direc[1:]))
		if err != nil {
			fmt.Println(err)
		}
		facing = getDirection(facing, turn)
		switch facing {
		case "N":
			hq_location.x += distance
		case "S":
			hq_location.x -= distance
		case "E":
			hq_location.y += distance
		case "W":
			hq_location.y -= distance
		}
	}
	return blocksAway(hq_location.x, hq_location.y)
}

func part2(directions []string) int {
	loc := Coordinate{0, 0}
	facing := "N"
	visited := make(map[Coordinate]bool)
	for _, direc := range directions {
		turn := string(direc[0])
		distance, err := strconv.Atoi(string(direc[1:]))
		if err != nil {
			fmt.Println(err)
		}
		facing = getDirection(facing, turn)
		for i := 0; i < distance; i++ {
			switch facing {
			case "N":
				loc.x += 1
			case "S":
				loc.x -= 1
			case "E":
				loc.y += 1
			case "W":
				loc.y -= 1
			}
			if visited[loc] == false {
				visited[loc] = true
			} else {
				return blocksAway(loc.x, loc.y)
			}
		}
	}
	return 0
}

func main() {
	path := "puzzle_input.txt"
	file, err := ioutil.ReadFile(path)
	if err != nil {
		fmt.Println("ERROR: could not open file")
		fmt.Println(err)
	}
	r, err := regexp.Compile("L\\d+|R\\d+")
	if err != nil {
		fmt.Println("ERROR: ", err)
	}

	directions := r.FindAllString(string(file), 176)

	fmt.Println("Part 1:", part1(directions))
	fmt.Println("Part 2:", part2(directions))
}

File: 2016/Day_1/day1.java
import java.io.File;
import java.util.Scanner;
import java.util.ArrayList;
import java.lang.Math;

public class day1 {
    public static void main (String[] args) {
        String path = "puzzle_input.txt";
        File myFile = new File(path);
        ArrayList<String> directions = new ArrayList<>();
        try {
            Scanner scanner = new Scanner(myFile);
            scanner.useDelimiter(", ");
            while (scanner.hasNext()) {
                directions.add(scanner.next());
            }
            scanner.close();
        } catch (Exception e) {
            System.out.println("ERROR:");
            e.printStackTrace();
            System.exit(0);
        }

        System.out.println("Part 1: " + part1(directions));
        System.out.println("Part 2: " + part2(directions));
    }

    static char getDirection(char facing, char turning) {
        char[] compass = new char[]{'W', 'N', 'E', 'S'};
        int facing_index = 0;
        for (int i = 0; i < compass.length; i++) {
            if (facing == compass[i]) {
                facing_index = i;
            }
        }
        if (turning == 'L') {
            if (facing_index - 1 >= 0) {
               return compass[facing_index - 1];
            } else {
                return compass[compass.length - 1];
            }
        } else {
            if (facing_index + 1 < compass.length) {
                return compass[facing_index + 1];
            } else {
                return compass[0];
            }
        }
    }

    static int blocksAway(int x, int y) {
        return Math.abs(x) + Math.abs(y);
    }

    static int part1(ArrayList<String> directions) {
        Coordinates hq_location = new Coordinates(0, 0);
        char facing = 'N';
        for (String direc: directions) {
            char turn = direc.charAt(0);
            int distance = Integer.parseInt(direc.substring(1));
            facing = getDirection(facing, turn);
            if (facing == 'N') {
                hq_location.incY(distance);
            } else if (facing == 'S') {
                hq_location.decY(distance);
            } else if (facing == 'E') {
                hq_location.incX(distance);
            } else {
                hq_location.decX(distance);
            }
        }   
        return blocksAway(hq_location.getX(), hq_location.getY());
    }

    static int part2(ArrayList<String> directions) {
        Coordinates loc = new Coordinates(0,0);
        char facing = 'N';
        ArrayList<ArrayList<Integer>> visited = new ArrayList<ArrayList<Integer>>();
        for (String direc: directions) {
            char turn = direc.charAt(0);
            int distance = Integer.parseInt(direc.substring(1));
            facing = getDirection(facing, turn);
            for (int i = 0; i < distance; i++) {
                if (facing == 'N') {
                    loc.incY(1);
                } else if (facing == 'S') {
                    loc.decY(1);
                } else if (facing == 'E') {
                    loc.incX(1);
                } else {
                    loc.decX(1);
                }
                ArrayList<Integer> current = new ArrayList<>();
                current.add(loc.getX());
                current.add(loc.getY());
                if (visited.contains(current)) {
                    return blocksAway(loc.getX(), loc.getY());
                } else {
                    visited.add(current);
                }
            }
        }
        return 0;
    }
}
